warl_proposal/csr_template_draft1.yaml

```diff
@@ -0,0 +1,156 @@
+        legal: [list of warl-string]
+        wr_illegal: [list of warl-string]           
+  rv64:                                 # this node and its subsequent fields can exist if [M/S/U]XL value can be 2
+    exists: <boolean>                   # indicates if this field exists in rv64 mode or not                          
+  rv128:                                # this node and its subsequent fields can exist if [M/S/U]XL value can be 2
+    exists: <boolean>                   # indicates if this field exists in rv64 mode or not                          
+##############################################################################################################
+
+############################################ Examples ############################################
+
+## example of csr with fields. The following is what the user will provide as input to riscv-config
+
+#User Input
+mtvec:                                
+  reset-val: 0x00000001 
+  rv32:                                 
+    exists: True                         
+    mode:                    
+      reset-val: 0x1
+      shadow: none
+      type:                             
+        WARL:
+          dependency_fields: []
+          legal: 
+            - "[] -> mode[1:0] in [0x0:0x1] # Range of 0 to 1 (inclusive)"
+          wr_illegal:
+            - "[] -> Unchanged"
+  base:
+      reset-val: 0x0
+      shadow: none
+      type:                             
+        WARL: 
+          dependency_fields: [mtvec::mode]
+          legal:
+            - "[0] -> base[29:0] in [0x20000000, 0x20004000]"             # can take only 2 fixed values in direct mode.
+            - "[1] -> base[29:6] in [0x000000:0xF00000] base[5:0] in [0x00]" # 256 byte aligned values only in vectored mode.
+          wr_illegal:
+            - "[0] -> Unchanged"
+            - "[1] wr_val in [0x2000000:0x4000000] -> 0x2000000"
+            - "[1] wr_val in [0x4000001:0x3FFFFFFF] -> Unchanged"
+  rv64:                                 
+    exists: False                   
+  rv128:                                
+    exists: False
+
+#The following is what the riscv-config will output after performing relevant checks on the above user-input
+#Checker output
+mtvec:                                
+  description: Machine trap vector base address
+  address: 0x305         
+  priv_mode: M
+  reset-val: 0x00000001 
+  rv32:                                 
+    exists: True                        
+    fields:                          
+      - mode
+      - base
+      - wpri                            # there is not wpri field in the mtvec 
+    mode:                   
+      description: trap handling mode
+      shadow: none
+      msb: 1
+      lsb: 0
+      implemented: True 
+      reset-val: 0x1
+      type:                             
+        WARL:
+          dependency_fields: []
+          legal: 
+            - "[] -> mode[1:0] in [0x0:0x1] # Range of 0 to 1 (inclusive)"
```

Comment:
Can "[] ->"  be removed if there's no dependency?

Reply:
Wouldn't having it make parsing easy? 
I am assuming one would use regex to extract parts from the string.

Reply:
I think it should not add much complexity to parser. User can have 2 regx, one with  "->", one w/o.   The one without "->" could be interpreted as no dependency.

Reply:
By the way, I think the complexity should be hidden from the user. A standard parser should be provided so that user only needs to deal with parsed information.

Reply:
Yes, we do intend to provide the parser in the package, but what about other implementations which do not use python to interpret the yaml. It adds a lot of complexity for them.

Reply:
I'd be OK to leave it like this, just a nice to have feature to improve readability and I believe the complexity is low and hidden from the user. My original thinking is that this format and the checker/parser script should be considered as a whole package. If user wants to process the YAML with other language, sounds to me it's a duplicated effort.

Reply:
One more note, I think there are two types of users for this YAML template:
1) Tool owner, like me, ovpsim, sail-riscv, compliance group etc.
2) RISC-V processor designer, verification engineer etc.
The parser complexity is exposed to (1) and can be reduced by providing a standard parser. The format complexity is exposed to (2). I'd actually like to call (1) as developer and (2) as user. All my suggestion is to make (2) life easier by adding some complexity to (1) :)